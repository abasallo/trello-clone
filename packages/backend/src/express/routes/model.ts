import HttpStatus from 'http-status-codes'

import { model as sequelizeModel } from '../../app'

import { Router } from 'express'

const router = Router()

router.get('/:id', async (req, res) => {
  const model = await (await sequelizeModel).Model.findOne({ where: { id: req.params.id } })
  if (!model) res.status(HttpStatus.NOT_FOUND)
  res.send(model)
})

router.post('/', async (req, res) =>
  res.send(
    await (
      await sequelizeModel
    ).Model.create({
      makeId: req.body.makeId,
      name: req.body.name,
      similarSoundingWordToNameParagraph: 'fewWordsSoundingLikeHardcodedResponse'
    })
  )
)

router.put('/', async (req, res) => {
  const model = await (await sequelizeModel).Model.findOne({ where: { id: req.body.id } })
  if (model) {
    ;(await sequelizeModel).Model.update(
      { makeId: req.body.makeId, name: req.body.name, similarSoundingWordToNameParagraph: '' },
      { where: { id: req.body.id } }
    )
    res.status(HttpStatus.OK)
  } else res.status(HttpStatus.NOT_FOUND)
  return res.send()
})

router.delete('/:id', async (req, res) => {
  const model = await (await sequelizeModel).Model.findOne({ where: { id: req.params.id } })
  if (model) {
    model.destroy()
    res.status(HttpStatus.OK)
  } else res.status(HttpStatus.NOT_FOUND)
  return res.send()
})

export default router
