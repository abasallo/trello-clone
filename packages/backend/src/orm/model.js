import Sequelize from 'sequelize'

const User = (sequelize) =>
  sequelize.define('user', {
    names: { type: Sequelize.STRING },
    surnames: { type: Sequelize.STRING },
    email: { type: Sequelize.STRING }
  })

const Make = (sequelize) =>
  sequelize.define('make', {
    id: { type: Sequelize.UUID, defaultValue: Sequelize.UUIDV4, primaryKey: true },
    name: { type: Sequelize.STRING }
  })

const Model = (sequelize) =>
  sequelize.define('model', {
    id: { type: Sequelize.UUID, defaultValue: Sequelize.UUIDV4, primaryKey: true },
    makeId: { type: Sequelize.STRING },
    name: { type: Sequelize.STRING },
    similarSoundingWordToNameParagraph: { type: Sequelize.STRING }
  })

const Car = (sequelize) =>
  sequelize.define('car', {
    id: { type: Sequelize.UUID, defaultValue: Sequelize.UUIDV4, primaryKey: true },
    modelId: { type: Sequelize.STRING },
    colour: { type: Sequelize.STRING },
    year: { type: Sequelize.STRING }
  })

Car.associate = () => Car.belongsTo(Model, { foreignKey: 'modelId', as: 'company' })
Model.associate = () => Model.belongsTo(Make, { foreignKey: 'makeId', as: 'company' })

export const initializeModel = (sequelize) => ({
  User: User(sequelize),
  Car: Car(sequelize),
  Model: Model(sequelize),
  Make: Make(sequelize)
})
