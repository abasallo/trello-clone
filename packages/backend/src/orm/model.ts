import Sequelize from 'sequelize'

// TODO - Proper typings and remove ts-ignore
// @ts-ignore
const User = (sequelize) =>
  sequelize.define('user', {
    names: { type: Sequelize.STRING },
    surnames: { type: Sequelize.STRING },
    email: { type: Sequelize.STRING }
  })

// TODO - Proper typings and remove ts-ignore
// @ts-ignore
const Make = (sequelize) =>
  sequelize.define('make', {
    id: { type: Sequelize.UUID, defaultValue: Sequelize.UUIDV4, primaryKey: true },
    name: { type: Sequelize.STRING }
  })

// TODO - Proper typings and remove ts-ignore
// @ts-ignore
const Model = (sequelize) =>
  sequelize.define('model', {
    id: { type: Sequelize.UUID, defaultValue: Sequelize.UUIDV4, primaryKey: true },
    makeId: { type: Sequelize.STRING },
    name: { type: Sequelize.STRING },
    similarSoundingWordToNameParagraph: { type: Sequelize.STRING }
  })

// TODO - Proper typings and remove ts-ignore
// @ts-ignore
const Car = (sequelize) =>
  sequelize.define('car', {
    id: { type: Sequelize.UUID, defaultValue: Sequelize.UUIDV4, primaryKey: true },
    modelId: { type: Sequelize.STRING },
    colour: { type: Sequelize.STRING },
    year: { type: Sequelize.STRING }
  })

// TODO - Proper typings and remove ts-ignore
// @ts-ignore
Car.associate = () => Car.belongsTo(Model, { foreignKey: 'modelId', as: 'company' })
// TODO - Proper typings and remove ts-ignore
// @ts-ignore
Model.associate = () => Model.belongsTo(Make, { foreignKey: 'makeId', as: 'company' })

// TODO - Proper typings and remove ts-ignore
// @ts-ignore
export const initializeModel = (sequelize) => ({
  User: User(sequelize),
  Car: Car(sequelize),
  Model: Model(sequelize),
  Make: Make(sequelize)
})
