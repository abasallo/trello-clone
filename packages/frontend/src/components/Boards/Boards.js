import React from 'react'

import PropTypes from 'prop-types'

import AddIcon from '@mui/icons-material/Add'

import { Grid } from '@material-ui/core'

import { AddButton, Board, BoardsGrid, Container } from './Boards.styled.components'

const Boards = (props) => {
  return (
    <Container>
      <BoardsGrid container spacing={2}>
        {props.boards.map((board) => (
          <Grid item xs={12} sm={6} md={4} lg={3} key={board.id}>
            <Board>{board.name}</Board>
          </Grid>
        ))}
      </BoardsGrid>
      <AddButton size="medium" color="secondary" aria-label="add">
        <AddIcon />
      </AddButton>
    </Container>
  )
}

Boards.propTypes = {
  boards: PropTypes.object
}

export default Boards
