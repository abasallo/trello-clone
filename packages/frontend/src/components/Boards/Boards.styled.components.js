import styled from 'styled-components'

import { Card, Fab, Grid } from '@material-ui/core'

import MaterialUIContainer from '@material-ui/core/Container'

export const Container = styled(MaterialUIContainer)`
  display: flex;
  flex-direction: column;
`

export const BoardsGrid = styled(Grid)`
  flex-grow: 1;
`
export const Board = styled(Card)`
  padding: 10px;
  min-height: 100px;
  text-align: center;
`

export const AddButton = styled(Fab)`
  margin-left: auto;
  margin-top: 10px;
`
