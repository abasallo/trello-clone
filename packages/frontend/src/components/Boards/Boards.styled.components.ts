import styled from 'styled-components'

import { Card, CardActions, Fab, Grid } from '@material-ui/core'

import MaterialUIContainer from '@material-ui/core/Container'

export const Container = styled(MaterialUIContainer)`
  display: flex;
  flex-direction: column;
`

export const BoardsGrid = styled(Grid)`
  flex-grow: 1;
`
export const BoardComponent = styled(Card)`
  padding: 10px;
  margin: 10px;
  min-height: 100px;
  text-align: center;
  background-color: ${(props) => props.color || '#000000'};
`

export const BoardActions = styled(CardActions)`
  text-align: right;
`

export const AddButton = styled(Fab)`
  margin-left: auto;
  margin-top: 10px;
`
