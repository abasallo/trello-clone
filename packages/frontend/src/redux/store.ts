import {configureStore} from '@reduxjs/toolkit'

import userReducer from './slices/user.slice'
import boardsReducer from './slices/boards.slice'

import {fetchBoardsAsyncThunk} from './thunks/boards.thunks'

export const reducer =  {
    user: userReducer,
    boards: boardsReducer,
}

export const store = configureStore({reducer})

store.dispatch(fetchBoardsAsyncThunk())

export type RootState = ReturnType<typeof store.getState>
export type AppDispatch = typeof store.dispatch
