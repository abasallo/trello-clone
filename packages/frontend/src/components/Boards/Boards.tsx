import React, {useState} from 'react'

import AddIcon from '@mui/icons-material/Add'
import DeleteIcon from '@mui/icons-material/Delete'
import EditIcon from '@mui/icons-material/Edit'

import {Chip, Grid, IconButton, TextField} from '@material-ui/core'

import {AddButton, BoardComponent, BoardActions, BoardsGrid, Container} from './Boards.styled.components'

import {useAppDispatch, useAppSelector} from '../../redux/hooks'
import {AppDispatch} from '../../redux/store'

import {addBoardAsyncThunk, deleteBoardAsyncThunk, updateBoardAsyncThunk} from '../../redux/thunks/boards.thunks'

import {Board} from 'trello-clone-shared/src/model/board.model'

const addButtonOnClick = (dispatch: AppDispatch): void => {
    dispatch(addBoardAsyncThunk())
}
const updateButtonOnClick = (dispatch: AppDispatch, board: Board): void => {
    dispatch(updateBoardAsyncThunk(board))
}

const deleteButtonOnClick = (dispatch: AppDispatch, id?: number): void => {
    if (id) dispatch(deleteBoardAsyncThunk(id))
}

const Boards = () => {

    const [editMode, setEditMode] = useState(false)

    const state = useAppSelector(state => state)
    const dispatch: AppDispatch = useAppDispatch()

    const boards: Board[] = state.boards || []

    return (
        <Container>
            <BoardsGrid container spacing={2}>
                {boards.map((board) => (
                    <Grid item xs={12} sm={6} md={4} lg={3} key={board.id}>
                        <BoardComponent color={board.colour}>
                            {
                                editMode
                                    ? <TextField id="standard-basic" label="Standard" variant="standard"
                                                 defaultValue={board.name}
                                                 onBlur={(e) => {
                                                     updateButtonOnClick(dispatch, {
                                                         id: board.id,
                                                         name: e.target.value
                                                     })
                                                     setEditMode(false)
                                                 }}
                                    />
                                    : <Chip label={board.name} variant="outlined" onClick={() => setEditMode(true)}/>
                            }

                            <BoardActions disableSpacing>
                                <IconButton aria-label="edit" onClick={() => setEditMode(true)}>
                                    <EditIcon/>
                                </IconButton>
                                <IconButton aria-label="delete" onClick={() => deleteButtonOnClick(dispatch, board.id)}>
                                    <DeleteIcon/>
                                </IconButton>
                            </BoardActions>
                        </BoardComponent>
                    </Grid>
                ))}
            </BoardsGrid>
            <AddButton size="medium" color="secondary" aria-label="add" onClick={() => addButtonOnClick(dispatch)}>
                <AddIcon/>
            </AddButton>
        </Container>
    )
}

export default Boards
