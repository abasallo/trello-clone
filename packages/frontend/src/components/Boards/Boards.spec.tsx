import React from 'react'
import renderer from 'react-test-renderer'

import {Provider} from 'react-redux'
import {configureStore} from '@reduxjs/toolkit'

import {reducer} from '../../redux/store'

import {User} from 'trello-clone-shared/src/model/user.model'
import {Board} from 'trello-clone-shared/src/model/board.model'

import Boards from './Boards'

// TODO:: Extract this
const user: User = {names: 'names', surnames: 'surnames', email: 'email'}
const boards: Board[] = [{id: 1, name: 'name-1'}, {id: 2, name: 'name-2'}]

// TODO:: Consider adding an integration test with actual mocked fetchers
export const store = configureStore({
    reducer,
    preloadedState: {user, boards}
})

test('Renders properly', () => expect(renderer.create(
    <Provider store={store}>
        <Boards/>
    </Provider>
).toJSON()).toMatchSnapshot())
