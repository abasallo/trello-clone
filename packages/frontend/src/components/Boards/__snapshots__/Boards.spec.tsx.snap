// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders properly 1`] = `
<div
  className="MuiContainer-root sc-bcXHqe bRLOeJ MuiContainer-maxWidthLg"
>
  <div
    className="MuiGrid-root sc-gswNZR fsLIpU MuiGrid-container MuiGrid-spacing-xs-2"
  >
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root MuiCard-root sc-dkrFOg bRFqtD MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          className="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-clickable"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="MuiChip-label"
          >
            name-1
          </span>
        </div>
        <div
          className="MuiCardActions-root sc-hLBbgP YyCSN"
        >
          <button
            aria-label="edit"
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-label="delete"
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
    <div
      className="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-4 MuiGrid-grid-lg-3"
    >
      <div
        className="MuiPaper-root MuiCard-root sc-dkrFOg bRFqtD MuiPaper-elevation1 MuiPaper-rounded"
      >
        <div
          className="MuiButtonBase-root MuiChip-root MuiChip-outlined MuiChip-clickable"
          onBlur={[Function]}
          onClick={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          role="button"
          tabIndex={0}
        >
          <span
            className="MuiChip-label"
          >
            name-2
          </span>
        </div>
        <div
          className="MuiCardActions-root sc-hLBbgP YyCSN"
        >
          <button
            aria-label="edit"
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="EditIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M3 17.25V21h3.75L17.81 9.94l-3.75-3.75L3 17.25zM20.71 7.04c.39-.39.39-1.02 0-1.41l-2.34-2.34a.9959.9959 0 0 0-1.41 0l-1.83 1.83 3.75 3.75 1.83-1.83z"
                />
              </svg>
            </span>
          </button>
          <button
            aria-label="delete"
            className="MuiButtonBase-root MuiIconButton-root"
            disabled={false}
            onBlur={[Function]}
            onClick={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="button"
          >
            <span
              className="MuiIconButton-label"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="DeleteIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M6 19c0 1.1.9 2 2 2h8c1.1 0 2-.9 2-2V7H6v12zM19 4h-3.5l-1-1h-5l-1 1H5v2h14V4z"
                />
              </svg>
            </span>
          </button>
        </div>
      </div>
    </div>
  </div>
  <button
    aria-label="add"
    className="MuiButtonBase-root MuiFab-root sc-eDvSVe gPAfeZ MuiFab-sizeMedium MuiFab-secondary"
    disabled={false}
    onBlur={[Function]}
    onClick={[Function]}
    onDragLeave={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={0}
    type="button"
  >
    <span
      className="MuiFab-label"
    >
      <svg
        aria-hidden={true}
        className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
        data-testid="AddIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
        />
      </svg>
    </span>
  </button>
</div>
`;
